docs/reverseproxy.md

```diff
@@ -0,0 +1,157 @@
+- At Level 3, Envoy has an identical setup as the proxy at Level 2, forwarding all traffic up to Level 4.
+- At Level 4, there is a specific set of listeners and destination clusters. For each authorized domain name, the proxy has a designated destination cluster that directs the traffic to the intended public domain name.
+
+> It worth noting that at Level 4, the Envoy Pod has custom network setup that does not use the rewritten domains by the local CoreDNS in Kubernetes. This guarantees that local address translation within the cluster forwards a set of domain names to the Envoy proxy Endpoint on the same Level 4 cluster. However, Envoy Pod itself resolves public addresses with the Azure or public DNS, overwriting its local DNS resolution to prevent an infinite redirection loop of DNS requests to the proxy itself. This implementation can be seen in the `hostNetwork` and `dnsPolicy` settings of the Envoy deployment manifest (in Helm).
+
+## DNS Rewriting with CoreDNS
+
+Within each cluster, regardless of the cluster's ability to directly access the Internet, CoreDNS is adjusted to modify a predefined set of domain names. This set includes all the fully qualified domain names specified in the Azure Arc Networking Requirements document, as well as some domain names necessary for downloading Helm charts, using GitOps, and accessing container images.
+
+Kubernetes has built-in DNS resolution capabilities for services and pods, as described in this relevant source [document](https://kubernetes.io/docs/concepts/services-networking/dns-pod-service/). CoreDNS is now the most used DNS service for Kubernetes and is available in Azure AKS.
+
+Each service or pod in the Kubernetes cluster is automatically provided with local DNS resolution injection. This enables the resolution of external domain names as well as inter-service domain name resolution.
+
+For modifying some of the domain names that must pass through the Envoy proxy, the CoreDNS plugin model is employed. Customization can be done using the `hosts` and custom `forward` servers.
+
+Upon a successful installation of the Envoy proxy, a load balancer service endpoint listening on port 443 is generated. This internal endpoint is assigned to a set of domain names to be overwritten via the `hosts` and `rewrite` plugin model for CoreDNS. In Azure AKS, this customization is possible but limited, and it can be accomplished through the ConfigMap `coredns-custom`. Please see [Customize CoreDNS with Azure Kubernetes Service](https://learn.microsoft.com/en-us/azure/aks/coredns-custom) for more information.
+
+## Virtual Network Configuration
+
+### Inbound (Ingress) Network Configuration
+
+At the networking level within the Virtual Network, all traffic from lower layers in a peered VNET is allowed through. This can further be locked down and in this case NSGs should permit traffic inbound to upper levels for ports 443 (HTTPS), 8883 (MQTT), 8084 (for Arc) and 53 (custom DNS).
```

Comment:
Not sure, did add a comment re allowing all FQDN resolutions via DNS at top level being less ideal but allowed in sample?

Reply:
Do you mean with L4 having full Internet access and using the default Azure DNS resolution? There is a note about potential usage of an enterprise firewall vs full Internet access, but DNS resolution for public FQDN will typically be at the enterprise's level DNS which may be pointing to a public DNS. 

There is a mention of usage of DNSMasq as a non-production sample, but I think you mean something else with your comment. 

Where do you suggest adding the note?

Reply:
Yeah, I believe in the current implementation, all FQDNs will be resolved from all layers regardless of whether those are permitted to connect via proxy or not (that's impression I had from our discussion but please correct). A note highlighting this as a potential breach of Principle of Least Privilege might be useful so users can lock DNS to only permissible FQDNs?

Reply:
Yes you are right, got it now and remember our discussion. I've added a paragraph about default resolution for all FQDNS and how this should be prevented in production scenario.